userspace/libscap/scap_open.h

```diff
@@ -31,79 +31,125 @@ limitations under the License.
-}scap_open_args;
+	typedef struct
+	{
+		bool ppm_sc[PPM_SC_MAX];
+	} interesting_ppm_sc_set;
+
+	typedef enum
+	{
+		UNKNOWN_ENGINE = 0,
+		KMOD_ENGINE = 1,
+		BPF_ENGINE = 2,
+		UDIG_ENGINE = 3,
+		NODRIVER_ENGINE = 4,
+		SAVEFILE_ENGINE = 5,
+		PLUGIN_ENGINE = 6,
+		GVISOR_ENGINE = 7,
+		MODERN_BPF_ENGINE = 8,
+		TEST_INPUT_ENGINE = 9,
+	} scap_engine_t;
+
+	typedef struct scap_open_args
+	{
+		scap_engine_t engine;
+		scap_mode_t mode;
+		proc_entry_callback proc_callback;			 ///< Callback to be invoked for each thread/fd that is extracted from /proc, or NULL if no callback is needed.
+		void* proc_callback_context;				 ///< Opaque pointer that will be included in the calls to proc_callback. Ignored if proc_callback is NULL.
+		bool import_users;					 ///< true if the user list should be created when opening the capture.
+		const char* suppressed_comms[SCAP_MAX_SUPPRESSED_COMMS]; ///< A list of processes (comm) for which no
+									 // events should be returned, with a trailing NULL value.
+									 // You can provide additional comm
+									 // values via scap_suppress_events_comm().
+		interesting_ppm_sc_set ppm_sc_of_interest; ///< syscall of interest
+
+		union
+		{
+			struct
+			{
+				uint64_t single_buffer_dim; ///<  dim of a single shared buffer. Usually, we have one buffer for every online CPU.
+			} kmod_args;
+
+			struct
+			{
+				uint64_t single_buffer_dim; ///<  dim of a single shared buffer. Usually, we have one buffer for every online CPU.
+				const char* bpf_probe;	    ///<  The path to the BPF probe object file.
+			} bpf_args;
+
+			struct
+			{
+				uint64_t single_buffer_dim; ///<  dim of a single shared buffer. Usually, we have one buffer for every online CPU.
+			} udig_args;
+
+			struct
+			{
+				int fd;		       ///< If non-zero, will be used instead of fname.
+				const char* fname;     ///< The name of the file to open.
+				uint64_t start_offset; ///< Used to start reading a capture file from an arbitrary offset. This is leveraged when opening merged files.
+			} scap_file_args;
+
+			struct
+			{
+				scap_source_plugin* input_plugin; ///< use this to configure a source plugin that will produce the events for this capture
+				char* input_plugin_params;	  ///< optional parameters string for the source plugin pointed by src_plugin
+			} plugin_args;
+
+			struct
+			{
+				const char* gvisor_root_path;	///< When using gvisor, the root path used by runsc commands
+				const char* gvisor_config_path; ///< When using gvisor, the path to the configuration file
+			} gvisor_args;
+
+			struct
+			{
+				uint64_t single_buffer_dim; ///<  dim of a single shared buffer. Usually, we have one buffer for every online CPU.
+			} modern_bpf_args;
+
+			struct
+			{
+				scap_test_input_data* test_input_data; ///<  only used for testing scap consumers by supplying arbitrary test data.
+			} test_input_args;
+		};
+
+	} scap_open_args;
```

Comment:
I'm not sure I'm a fan of this either. It makes libscap intimately familiar with the implementation of all engines.

I think I'd rather have a `void* engine_args` that gets passed to the individual engine which then casts it to whatever type it wants.

The consumer could then look like e.g.

```
#include <libscap/engine/kmod/kmod_public.h>

// ...

struct scap_kmod_engine_params params = { ... };

scap_open("kmod", ..., &params);

// ...
```

Yes, this exposes the engine to the consumer instead, but:
1. the consumer must know the engine parameters anyway
2. libscap does nothing with these params, just shuffles them around

If we continue down this path, libscap itself might just be gone eventually, except as an API that all engines must conform to and TBH I'm not sure it's a bad thing (the non-engine parts of libscap either provide very little value or belong elsewhere).

Reply:
Yeah the `union` is one of the 2 choices, at the beginning, I implemented it through opaque pointers but to avoid all the casting I have chosen a `union` approach, but in the end, yes, probably we have more advantages from an opaque pointer approach, I can change it :)

Reply:
In C (not C++, sorry gvisor), you can simply assign a void* to a foo* without a cast.

Reply:
Uh, that's true I forgot it :)